### ISSUE TITLE
Safe exception handling with Warp?

### ISSUE THREAD

**Issue by mkscrg** [yesodweb/wai#555]
Is it safe to catch all exceptions at the top level of a `wai`/`warp` app? [1]
- If safe and correct, let's document this on `warp`. Catching all exceptions is usually an antipattern, etc etc
- If not safe, is it reasonable to use something like `enclosed-exceptions`? I'd worry about that affecting performance, though I haven't tried it.

FWIW, this question came up because we'd like access to the `Request` when constructing a 500 response. `warp`'s current `onExceptionResponse` doesn't pass the `Request`.

[1] Like so:
```haskell
myApp :: Application
myApp req respond = respond . response500 req `handle` theRestOfMyApp

response500 :: Request -> SomeException -> Response
response500 req someEx = -- ...
```

**Comment by snoyberg** [comment]
@kazu-yamamoto do you have thoughts on this? I'm not sure how the HTTP2 code affects things. AFAIK, it's safe to catch exceptions, though if you catch a `TimeoutThread` exception you should probably rethrow it.

**Comment by kazu-yamamoto** [comment]
I believe all exception should be caught by the code to run Warp regardless of HTTP/1 and HTTP/2.
If it is not the case, it's a bug of Warp.

**Comment by snoyberg** [comment]
Addressed by #557
